### PR TITLE
CCMath 0.3.0 Release Candidate

### LINKED ISSUE #135: [BUG]: constexpr math functions fail to compile on domain/special inputs
### Due Diligence

- [x] I have searched for similar issues before starting this bug report

### Describe the bug

Several functions call the non-constexpr set_errno_if_required on their domain-error/special path without an is_constant_evaluated() guard, so evaluating them in a constant expression is ill-formed. 

Guarding is inconsistent: sqrt(-1), fmod(1,0), and scalbn(inf,2) compile fine.

### Steps to reproduce

```cpp
#include <ccmath/ccmath.hpp>

// each of these fails to compile
constexpr double a = ccm::log(0.0);
constexpr double b = ccm::log(-1.0);
constexpr double c = ccm::acos(2.0);
constexpr double d = ccm::pow(-1.0, 0.5);
constexpr double e = ccm::tgamma(0.0);
constexpr int f = ccm::ilogb(0.0);

int main() {}
```

Error: non-constexpr function 'set_errno_if_required' cannot be used in a constant expression.

### Expected behavior

All should compile as constant expressions and return the correct special value (for example log(0) = -inf), raising no errno at compile time.

### Compiler version and type

Apple clang 17.0.0 (clang-1700.4.4.1)

### Operating System

_No response_

### Screenshots

Not applicable

### PR DESCRIPTION
**Pull request by Rinzii**
# CCMath 0.3.0 Release Candidate

This is the review PR for v0.3.0. The goal is to catch any remaining blockers before the
release is cut.

It puts the entire v0.2.0 to v0.3.0 change set in one diff. It is review-only and will not be
merged. The base branch `review-base/v0.2.0` is a frozen snapshot of the `v0.2.0` tag, so the
diff is exactly what landed since the last release. The content itself already lives on `main`
and on `release/0.3.x`.

How this review works: comments here are the review record. Fixes land as normal PRs into
`main`, after which `release/0.3.x` is fast-forwarded. Record your approval as a review here.
When review wraps up, the PR is closed and `review-base/v0.2.0` is deleted. Treat it as a diff
viewer and discussion thread.

The user-facing summary is in the release notes. What follows is for reviewers: what changed,
where the risk is, what is validated, and what would block the release.

> Heads up on size: 891 files, about +59.6k / -9.8k. Most of it is build, tooling, test, and
> doc work rather than runtime math kernels. The part that can produce silently wrong numbers
> though is small and it is listed below first.

## What counts as blocking

Reasons to hold the release. Mark these clearly so they stand out from nits:

- A numeric result that violates the documented contract or a release claim, in any rounding
  mode, on a supported path.
- Wrong dispatch or format selection (for example a `powl` call routed to the wrong kernel).
- A public API contract problem in the new `ext` or `fmanip` functions (signature, NaN and
  infinity behavior, constexpr-ness).
- A claim in the release notes the code does not actually support.
- A portability break on a supported compiler or standard mode (C++17 through C++26).

Most other items can be handled as nits or follow-ups unless they change the release risk.
Items under Known limitations are out of scope by design, not blockers, unless you can show
one breaks a supported path.

## Where to focus review

These are the surfaces where a mistake means wrong math, so they earn the closest reading.
Validation has been run on most of them (see Validation status), but that lowers the
uncertainty rather than removing the need for attention.

### Trigonometry (`double` `sin`, `cos`, `tan`)

- [sincos_payne_hanek.hpp](include/ccmath/math/trig/impl/sincos_payne_hanek.hpp) plus the
  `sin`, `cos`, and `tan` kernels.
- What to check: the Payne-Hanek large-argument reduction, the binary64 constant values, the
  quadrant selection, sign-of-zero handling, and the inverse-trig edge cases (`asin`, `acos`,
  `atan`, `atan2`). The earlier failure mode was large arguments falling outside the
  reduction, so large-input behavior is the thing to primarily probe.

### Power (`double` `pow`, `long double` `powl`)

- The double-double constant pairs in the `pow` path. The headline bug class fixed this
  release was hi and lo halves stored swapped, which is harmless under round-to-nearest but
  wrong in directed modes, so check each DD constant against its intended value and ordering.
- The new 80-bit `powl` kernel:
  [powl_ld80_kernel.hpp](include/ccmath/internal/math/generic/func/power/pow_impl/powl_ld80_kernel.hpp)
  and its generated
  [powl_ld80_tables.hpp](include/ccmath/internal/math/generic/func/power/pow_impl/powl_ld80_tables.hpp).
  It compiles only where `long double` is the 80-bit x87 extended format, so the `double` and
  IEEE binary128 `long double` platforms do not exercise it. This is preview quality (see
  Known limitations).
- SIMD `pow` paths: `pow_simd_impl.hpp` and `powf_simd_impl.hpp` in the same directory. These
  run only under runtime SIMD (not deterministic mode), so the check is that they agree with
  the scalar path on the same inputs.

### Configuration and dispatch

- [long_double_format.hpp](include/ccmath/internal/config/long_double_format.hpp) and
  [powl_policy.hpp](include/ccmath/internal/config/powl_policy.hpp) decide which `long double`
  path is taken. A wrong detection sends every `powl` call down the wrong kernel, so the
  format-detection logic is worth careful eyes.
- Deterministic mode (`CCMATH_ENABLE_DETERMINISTIC`): the claim is bit-identical `float` and
  `double` results across supported hardware. Worth confirming it forces the generic kernels, disables
  runtime SIMD, does not leak into a libm or compiler builtin on any path, and that constexpr
  and runtime agree. `long double` is not part of the deterministic claim.

### New public API

- The 15 new `ccm::ext` helpers under [ext/](include/ccmath/ext) and the standard `nextup`
  and `nextdown` under [fmanip/](include/ccmath/math/fmanip). Check signatures, NaN and
  infinity behavior, constexpr-ness, and that they build clean under `-Wconversion`. The `ext`
  helpers are opt-in behind `CCMATH_ENABLE_EXTENSIONS`, so also check the namespace placement
  and that they stay out of a default build.

### Low-level support

- [support/fp/](include/ccmath/internal/support/fp): `cast.hpp`, `except_value_utils.hpp`,
  `fma.hpp`. The changes here are small, but used very widely so mistake can easily propagate.

## Lower-risk areas (skim)

- `ccm::pp` SIMD layer under [runtime/pp/](include/ccmath/internal/math/runtime/pp): a
  standalone C++17 port of the C++26 `std::simd` interface. Not yet wired into dispatch, so it
  cannot change results this release. Review it on its own merits if you have the appetite,
  otherwise it can wait.
- Tests under [tests/](tests): new conformance, rigorous, and proof harnesses. The useful
  question here is coverage gaps against the kernels above.
- Fuzzing under [fuzzing/](fuzzing): libFuzzer targets per family.
- Tooling: asmlab under [tools/asmlab/](tools/asmlab), the showcase comparisons, and the
  benchmarks. Build and dev infrastructure, not part of library runtime behavior.
- Build systems: `meson.build`, `premake5.lua`, the CMake options, and the vendored-consumer
  integration test.
- CI under [.github/workflows/](.github/workflows): the C++23 and C++26 matrix and the
  per-PR clang-format gate.
- Docs and proofs: a partial `POW_PROOF.tex` (the full binary64 proof is still open) and the
  Sollya and Gappa scripts under `docs/approximating_functions`.

## Validation status

What has been validated, and how. This lowers the uncertainty on these areas, it does not put
them out of bounds for review:

- `pow`: CORE-MATH all-mode validation campaigns over the covered inputs, all four rounding
  modes. The campaign records live under `tests/rigorous/` (see `oracle_logs/` for the
  per-campaign summaries).
- `sin`, `cos`, `tan`: full-range validation campaigns, large arguments included. This is
  representative coverage across the range, not an exhaustive sweep of every binary64 value.
- `fma`: all-mode exact software fallback, with native runtime coverage validated on AArch64.
- Differential fuzzing across the basic, compare, exponential, float-manipulation, misc,
  nearest, power, and trig families.
- Builds under `-Werror` with aggressive warnings (`-Wconversion` and friends) across the CI
  matrix on Linux, macOS, and Windows.

Open validation work (not expected to be finished in this RC):

- The full binary64 `pow` proof. The writeup in the tree is partial.
- `powf` reduced-domain validation.
- Independent validation of the 80-bit `powl` kernel.
- Wider x86-64 `fma` validation.

## Known limitations and intended scope

These are disclosed in the release notes and are intentionally out of scope for v0.3.0. Call
one out if you think it should block the release, otherwise treat it as known debt:

- The 80-bit `powl` kernel is preview quality. Its generic fuzz lane is intentionally guarded
  off in [power_fuzz.cpp](fuzzing/src/math/power_fuzz.cpp) until the kernel is finalized.
- Deterministic mode is experimental.
- Hyperbolic and special functions are still empty headers.

## Building and validating locally

It is a header-only library, so consuming it is just an include. To exercise the validation
that matters for review:

```bash
cmake -S . -B build -DCCMATH_BUILD_RIGOROUS_TESTS=ON \
  -DCCMATH_ENABLE_WARNINGS_AS_ERRORS=ON -DCCMATH_ENABLE_AGGRESSIVE_WARNINGS=ON
cmake --build build --config Release
ctest --test-dir build --output-on-failure -C Release
```

On multi-config generators like MSVC the `--config` and `-C Release` flags matter, and they
are harmless elsewhere. Add `-DCCMATH_ENABLE_MPFR_TESTS=ON` and
`-DCCMATH_ENABLE_COREMATH_TESTS=ON` for the oracle-backed campaigns (these need MPFR and
CORE-MATH available, and the rigorous runs can take a while). Fuzzers build under Clang with
`-DCCMATH_BUILD_FUZZING=ON`. Run `tools/ensure_format.sh` for the format gate. The README
validation section and `tools/asmlab/README.md` cover the deeper tooling.

## Where review is most valuable

Highest value first: numeric correctness and edge cases in the kernels above, then the public
API shape of the new `ext` and `fmanip` functions, then test coverage gaps against those
kernels. Style and tooling nits are welcome but lowest priority for an RC this size.

### CODE REVIEW

File: include/ccmath/ext/abs_diff.hpp [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	/**
+	 * @brief Compute the absolute difference between two integral values without unsigned underflow.
+	 * @tparam T Type of the input and output.
+	 * @param lhs The first value.
+	 * @param rhs The second value.
+	 * @return The absolute difference between lhs and rhs.
+	 */
+	template <typename T, std::enable_if_t<std::is_integral_v<T> && !std::is_same_v<std::remove_cv_t<T>, bool>, bool> = true>
+	[[nodiscard]] constexpr T abs_diff(T lhs, T rhs) noexcept
+	{ return lhs > rhs ? static_cast<T>(lhs - rhs) : static_cast<T>(rhs - lhs); }
```
Comment (Eczbek) [inline include/ccmath/ext/abs_diff.hpp]: `abs_diff(INT_MAX, -1)` causes signed integer wrapping, perhaps it would make sense to return an unsigned type for integers?

**Review by Eczbek — CHANGES_REQUESTED** [review]

File: include/ccmath/ext/ceil_div.hpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#pragma once
+
+#include <type_traits>
+
+namespace ccm::ext
+{
+	/**
+	 * @brief Divide two integral values and round the result toward positive infinity.
+	 *
+	 * This function is intended for non-negative integral values.
```
Comment (Eczbek) [inline include/ccmath/ext/ceil_div.hpp]: Please add support for negative integral values! This may be used for reference: https://gist.github.com/eisenwave/2a7d7a4e74e99bbb513984107a6c63ef#file-impl_intdiv-hpp

File: include/ccmath/ext/chgsign.hpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+	 */
+	template <typename T, std::enable_if_t<std::is_floating_point_v<T>, bool> = true>
+	[[nodiscard]] constexpr T chgsign(T x, T y) noexcept
+	{
+		const T magnitude = ::ccm::abs(x);
+		const auto y_bits = ::ccm::support::bit_cast<::ccm::support::float_bits_t<T>>(y);
+
+		if ((y_bits & ::ccm::support::sign_mask_v<T>) != 0) { return -magnitude; }
+
+		return magnitude;
```
Comment (Eczbek) [inline include/ccmath/ext/chgsign.hpp]: Is this implementation better in some way than simply
```cpp
return (signbit(x) == signbit(y)) ? x : -x;
```
?

File: include/ccmath/ext/mix.hpp [hunk 4]
```diff
@@ -41,7 +41,5 @@ namespace ccm::ext
 	 */
 	template <typename TStart, typename TEnd, typename TAplha>
 	constexpr std::common_type_t<TStart, TEnd, TAplha> mix(TStart x, TEnd y, TAplha a) noexcept
-	{
-		return (x * (1 - a)) + (y * a);
-	}
+	{ return (x * (1 - a)) + (y * a); }
```
Comment (Eczbek) [inline include/ccmath/ext/mix.hpp]: This causes signed integer wrapping in many cases:
- `mix(0, 0, INT_MIN)` for `1 - a`
- `mix(0, INT_MAX, 2)` for `y * a`
- `mix(INT_MIN, 0, 2)` for `x * (1 - a)`

etc.

File: include/ccmath/ext/rcp.hpp [hunk 5]
```diff
@@ -22,7 +22,5 @@ namespace ccm::ext
 	 */
 	template <typename T, std::enable_if_t<std::is_floating_point_v<T>, bool> = true>
 	constexpr T rcp(T x) noexcept
-	{
-		return static_cast<T>(1) / x;
-	}
+	{ return static_cast<T>(1) / x; }
```
Comment (Eczbek) [inline include/ccmath/ext/rcp.hpp]: Casting of literals is inconsistent between files, sometimes `T(123)` and sometimes `static_cast<T>(123)`

File: include/ccmath/ext/remap.hpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+		 * @param dst_end End of the destination range.
+		 * @param x Value to remap.
+		 * @return The remapped value.
+		 */
+		template <typename T, std::enable_if_t<std::is_floating_point_v<T>, bool> = true>
+		[[nodiscard]] constexpr T remap(T src_start, T src_end, T dst_start, T dst_end, T x) noexcept
+		{
+			const T length = src_end - src_start;
+
+			if (length == T(0)) { return dst_start; }
```
Comment (Eczbek) [inline include/ccmath/ext/remap.hpp]: Does it make sense to compare floating-point values with `==`? Perhaps `approximately()` can be used instead. This might also apply to other files.

File: include/ccmath/ext/sign.hpp [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+{
+	/**
+	 * @brief Return the sign of a value.
+	 * @tparam T Type of the input and output.
+	 * @param value The value whose sign is queried.
+	 * @return -1 if value is negative, 0 if value is zero, otherwise 1.
+	 */
+	template <typename T, std::enable_if_t<std::is_arithmetic_v<T> && !std::is_same_v<T, bool>, bool> = true>
+	[[nodiscard]] constexpr T sign(T value) noexcept
+	{ return static_cast<T>((T(0) < value) - (value < T(0))); }
```
Comment (Eczbek) [inline include/ccmath/ext/sign.hpp]: Should this use `signbit()` for floating-point types?

File: include/ccmath/internal/config/compiler.hpp
Comment (Eczbek) [inline include/ccmath/internal/config/compiler.hpp:33]: It might be good to check if `_MSC_FULL_VER` defined, in which case
```cpp
#define CCMATH_COMPILER_MSVC_VER_MAJOR (_MSC_FULL_VER / 10000000)
#define CCMATH_COMPILER_MSVC_VER_MAJOR (_MSC_FULL_VER % 10000000 / 10000)
#define CCMATH_COMPILER_MSVC_VER_PATCH (_MSC_FULL_VER % 100000)
```
for completeness or something

File: include/ccmath/ext/fract.hpp [hunk 8]
```diff
@@ -24,7 +24,5 @@ namespace ccm::ext
 	 */
 	template <typename T, std::enable_if_t<std::is_floating_point_v<T>, bool> = true>
 	constexpr T fract(T x) noexcept
```
Comment (Eczbek) [inline include/ccmath/ext/fract.hpp]: Functions are inconsistently marked `[[nodiscard]]` between files.

File: include/ccmath/ext/abs_diff.hpp [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+	 * @param rhs The second value.
+	 * @return The absolute difference between lhs and rhs as the matching unsigned type.
+	 */
+	template <typename T, std::enable_if_t<std::is_integral_v<T> && !std::is_same_v<std::remove_cv_t<T>, bool>, bool> = true>
+	[[nodiscard]] constexpr std::make_unsigned_t<T> abs_diff(T lhs, T rhs) noexcept
+	{
+		using U = std::make_unsigned_t<T>;
+		return lhs > rhs ? static_cast<U>(static_cast<U>(lhs) - static_cast<U>(rhs)) : static_cast<U>(static_cast<U>(rhs) - static_cast<U>(lhs));
+	}
+} // namespace ccm::ext
```
Comment (Eczbek) [inline include/ccmath/ext/abs_diff.hpp]: Some files are inconsistently missing a trailing newline

File: include/ccmath/internal/math/common/basic/max.hpp [hunk 10]
```diff
@@ -30,8 +30,11 @@ namespace ccm
 	template <typename T>
 	constexpr T max(T x, T y) noexcept
 	{
-		if constexpr (ccm::builtin::has_constexpr_fmax<T>) { return ccm::builtin::fmax(x, y); }
-		else { return ccm::gen::max(x, y); }
+		if constexpr (ccm::builtin::has_constexpr_fmax<T>) { return ccm::builtin::fmax_ct(x, y); }
+		else
+		{
+			return ccm::gen::max(x, y);
+		}
```
Comment (Eczbek) [inline include/ccmath/internal/math/common/basic/max.hpp:38]: This formatting was changed and is now weird. This also happens in other files.

File: include/ccmath/internal/math/common/basic/fma.hpp [hunk 11]
```diff
@@ -119,12 +76,23 @@ namespace ccm
...
+	/**
+	 * @brief Fused multiply-add operation.
+	 * @param x Floating-point value.
+	 * @param y Floating-point value.
+	 * @param z Floating-point value.
+	 * @return If successful, returns the value of x * y + z as if calculated to infinite precision and rounded once to fit the result type (or, alternatively,
+	 * calculated as a single ternary floating-point operation).
+	 */
+	constexpr double fma(double x, double y, double z) noexcept
+	{ return ccm::fma<double>(x, y, z); }
```
Comment (Eczbek) [inline include/ccmath/internal/math/common/basic/fma.hpp]: Is explicitly passing a template argument necessary? This might be inconsistent between files.

File: include/ccmath/internal/math/common/nearest/floor.hpp [hunk 12]
```diff
@@ -24,27 +24,18 @@ namespace ccm
 		template <typename T>
 		constexpr T floor_pos_impl(T num) noexcept
 		{
-			// Calculate the maximum value that can be compared with 'num' for equality.
+			// At or above 1/epsilon, num is already integral (or indistinguishable from one) in T.
 			constexpr auto max_comparable_val = T(1) / std::numeric_limits<T>::epsilon();
```
Comment (Eczbek) [inline include/ccmath/internal/math/common/nearest/floor.hpp:27]: Division by zero for integer types. Not sure if this function is only intended for floating-point types?

File: include/ccmath/internal/math/generic/builtins/basic/fabs.hpp [hunk 13]
```diff
@@ -47,6 +47,21 @@
...
 #endif
 
+/// CCMATH_HAS_BUILTIN_ABS
+/// This is a macro that is defined if the compiler has constexpr __builtin functions for abs that allow static_assert
+///
+/// Compilers with Support:
+/// - GCC
+/// - Clang
+
+// TODO(IanP): Determine the lowest runtime compiler versions at some point.
```
Comment (Eczbek) [inline include/ccmath/internal/math/generic/builtins/basic/fabs.hpp:59]: The oldest compiler versions available on Compiler Explorer that seem to support `__builtin_fabs()` and friends are GCC 4.7.1 (due to not recognizing `-std=c++11` even earlier) and Clang 3.2 (oldest available): https://godbolt.org/z/xn5Wsa1j5 

But strangely Clang only recognizes `__builtin_abs()` since 20.1.0: https://godbolt.org/z/b5esjYh7T

File: include/ccmath/internal/math/generic/builtins/basic/fabs.hpp
Comment (Eczbek) [inline include/ccmath/internal/math/generic/builtins/basic/fabs.hpp]: Did you mean to `std::remove_cv_t<T>`? Checking `std::is_same_v<T, long long>` seems redundant because of the `std::is_integral_v<T>` check. Checking whether `T` is signed or unsigned seems redundant too, and is this intended to allow `bool`?

Also, it might be beneficial to make a helper:
```cpp
template <typename First, typename... Rest>
inline constexpr bool is_same_any = (... || std::is_same_v<First, Rest>);

// ...

template <typename T>
inline constexpr bool has_constexpr_abs = is_same_any<std::remove_cv_t<T>, float, double, long double> // ...
```

File: include/ccmath/internal/math/generic/func/basic/fmod_gen.hpp [hunk 14]
```diff
@@ -65,9 +65,7 @@ namespace ccm
 
 		template <typename T, typename U, typename TC = std::common_type_t<T, U>>
```
Comment (Eczbek) [inline include/ccmath/internal/math/generic/func/basic/fmod_gen.hpp]: Is `TC` intended to be customizable by the user?

File: include/ccmath/internal/math/runtime/pp/assume_aligned.hpp [hunk 15]
```diff
@@ -14,7 +14,8 @@ namespace ccm::pp
 	template <std::size_t Alignment>
 	CCM_ALWAYS_INLINE void * assume_aligned(void * ptr)
 	{
-		static_assert((Alignment & (Alignment - 1)) == 0, "Alignment must be a power of 2"); // TODO: Might remove this check not sure if it's necessary
+		// TODO: Decide whether the power-of-two alignment static_assert is needed.
+		static_assert((Alignment & (Alignment - 1)) == 0, "Alignment must be a power of 2");
```
Comment (Eczbek) [inline include/ccmath/internal/math/runtime/pp/assume_aligned.hpp]: Check that `Alignment` is nonzero

File: include/ccmath/internal/math/runtime/pp/flags.hpp [hunk 16]
```diff
@@ -11,168 +11,116 @@
...
+		struct flags_align_request : std::integral_constant<std::size_t, 0>
 		{
 		};
-
-		template <int L1, int L2>
-		struct Prefetch : LoadStoreTag
+		template <typename... Fs>
+		struct flags_align_request<simd_flags<Fs...>>
+			: std::integral_constant<std::size_t,
+									 ((std::size_t{ 0 } + ... + overalign_value<Fs>::value) != 0)
```
Comment (Eczbek) [inline include/ccmath/internal/math/runtime/pp/flags.hpp:56]: More inconsistent casting of literals, now with curly braces

File: include/ccmath/internal/support/fp/cast.hpp [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+#include "ccmath/internal/support/type_traits.hpp"
+#include "ccmath/internal/types/dyadic_float.hpp"
+
+#include <cfenv>
+#include <type_traits>
+
+namespace ccm::support::fp
+{
+	template <typename OutType, typename InType>
+	[[nodiscard]] constexpr std::enable_if_t<traits::ccm_is_floating_point_v<OutType> && traits::ccm_is_floating_point_v<InType>, OutType> cast(InType x)
```
Comment (Eczbek) [inline include/ccmath/internal/support/fp/cast.hpp:30]: Could this be generalized to also allow safely casting integers to floats and floats to integers?

File: include/ccmath/ext/mix.hpp (hunk 4 above)
Comment (Rinzii) [reply include/ccmath/ext/mix.hpp]: There is technical reasons to not support integers but also after some debate I've actually decided to completely remove mix due to its similarity to lerp and not being a proper representation of GLSL's mix function.

File: include/ccmath/ext/chgsign.hpp (hunk 3 above)
Comment (Rinzii) [reply include/ccmath/ext/chgsign.hpp]: After some deliberation I have decided to remove this due to it's similarity to std::copysign and how I think it's redundant here.

File: include/ccmath/internal/support/fp/cast.hpp (hunk 17 above)
Comment (Rinzii) [reply include/ccmath/internal/support/fp/cast.hpp:30]: fp::cast is FP-to-FP by design. It effectively ensures NaN/inf/subnormals have correct rounding via DyadicFloat. Accepting int <-> float is a massive can of worms with a ton of different correctness requirements and conflating them would only bloat a very hot and widely used primitive. So I would say I'd decline this recommendation in it's current state. Maybe down the line this might change, but if that were to be the case I'd likely be it's own separate function.

File: include/ccmath/internal/math/common/nearest/floor.hpp (hunk 12 above)
Comment (Rinzii) [reply include/ccmath/internal/math/common/nearest/floor.hpp:27]: This is incorrect as integrals have a dedicated overload at the earliest entry into the chain so this will never be instantiated with a non-integral.

File: include/ccmath/internal/math/runtime/pp/flags.hpp (hunk 16 above)
Comment (Rinzii) [reply include/ccmath/internal/math/runtime/pp/flags.hpp:56]: I could enforce a convention with an AST matcher but I'm not sure if it's worth the effort or not. Thoughts?

File: include/ccmath/internal/math/generic/func/basic/fmod_gen.hpp (hunk 14 above)
Comment (Rinzii) [reply include/ccmath/internal/math/generic/func/basic/fmod_gen.hpp]: It's an internal customization point and is only intended for developers, but yes.

File: include/ccmath/internal/math/runtime/pp/flags.hpp (hunk 16 above)
Comment (Rinzii) [reply include/ccmath/internal/math/runtime/pp/flags.hpp:56]: Based on discussion we had I think I'm going to mark this as low priority and WONTFIX.

File: include/ccmath/internal/config/compiler.hpp
Comment (Rinzii) [reply include/ccmath/internal/config/compiler.hpp:33]: This would be a nice enhancement but for this release is not required. I'd say open an issue on it though!

File: include/ccmath/ext/remap.hpp (hunk 6 above)
Comment (Rinzii) [reply include/ccmath/ext/remap.hpp]: It would not make much sense to switch off of using `==` as here it's usage is safe and correct as it's supposed to be guarding against division by zero and we need to be sure this is EXACT zero where using approximately would treat tiny but valid ranges as zero which would be incorrect. So this is a WONTFIX for me.

File: include/ccmath/ext/sign.hpp (hunk 7 above)
Comment (Rinzii) [reply include/ccmath/ext/sign.hpp]: I'd say no as I want to keep sign as a numeric signum rather than make it act like signbit.

File: include/ccmath/internal/math/common/basic/max.hpp (hunk 10 above)
Comment (Rinzii) [reply include/ccmath/internal/math/common/basic/max.hpp:38]: The formatting is considered valid under clang-format?

File: include/ccmath/internal/math/common/basic/fma.hpp (hunk 11 above)
Comment (Rinzii) [reply include/ccmath/internal/math/common/basic/fma.hpp]: In a general case? Probably not, but also keeping it does not hurt anything? So I'll keep this as a WONTFIX for now.

File: include/ccmath/ext/fract.hpp (hunk 8 above)
Comment (Rinzii) [reply include/ccmath/ext/fract.hpp]: Current policy is to have no `[[nodiscards]]` in the public api but they should be used properly throughout the internal code base. This should now be resolved with the latest pushes.

File: include/ccmath/ext/rcp.hpp (hunk 5 above)
Comment (Rinzii) [reply include/ccmath/ext/rcp.hpp]: CI should now enforce this across the board.

**Comment by Rinzii** [comment]
Ok, no more work is planned to be done here going forward besides patches and to address reviews.

**Comment by nekoflare** [comment]
I have an issue with this function:
```cpp
    /**
     * @brief Determines if a signed integer value is a power of two.
     *
     * @tparam T Type of the input.
     * @param n Value to check.
     * @return True if the input is a power of two, false otherwise.
     *
     * @note Prefer ccm::ext::safe::is_power_of_two for signed inputs to avoid signed
     * overflow on the minimum representable value.
     */
    template <typename T, std::enable_if_t<std::is_signed_v<T> && !std::is_same_v<T, bool>, bool> = true>
    constexpr bool is_power_of_two(T n) noexcept
    { return n && !(n & (n - T(1))); }
```
inside is_power_of_two.hpp file.
Because naturally negative numbers cannot be power of two.
Therefore you should add `n > T(0)`

Another issue being that for functions `delta_angle`, `lerp_angle`, and `move_towards_angle` inside the namespace `ccm::ext` you use degrees but everywhere else I can see you're using radians. What if user forgets to apply radians to degrees? I think you should add a suffix `_deg` to tell the user those functions operate on degrees rather than radians or rewrite those functions to use radians.

**Comment by Rinzii** [comment]
> I have an issue with this function:
> 
> ```c++
>     /**
>      * @brief Determines if a signed integer value is a power of two.
>      *
>      * @tparam T Type of the input.
>      * @param n Value to check.
>      * @return True if the input is a power of two, false otherwise.
>      *
>      * @note Prefer ccm::ext::safe::is_power_of_two for signed inputs to avoid signed
>      * overflow on the minimum representable value.
>      */
>     template <typename T, std::enable_if_t<std::is_signed_v<T> && !std::is_same_v<T, bool>, bool> = true>
>     constexpr bool is_power_of_two(T n) noexcept
>     { return n && !(n & (n - T(1))); }
> ```
> 
> inside is_power_of_two.hpp file. Because naturally negative numbers cannot be power of two. Therefore you should add `n > T(0)`
> 
> Another issue being that for functions `delta_angle`, `lerp_angle`, and `move_towards_angle` inside the namespace `ccm::ext` you use degrees but everywhere else I can see you're using radians. What if user forgets to apply radians to degrees? I think you should add a suffix `_deg` to tell the user those functions operate on degrees rather than radians or rewrite those functions to use radians.


Good catch! The original intention was for the unqualified signed overload to be the lean bit-twiddle version and for `ccm::ext::safe::is_power_of_two` to be the guarded one that handles negatives and the minimum value correctly, the thinking being that the default should stay as cheap as possible.

After some analysis though I compared the codegen for both and there is honestly no performance difference between them that I can find. They come out branchless at the same instruction count on clang, and gcc emits the same shape with a single compare, so guarding against the minimum value should costs nothing. Given that, there is no reason I can think of to keep a default that is wrong at the minimum value, so I am going to replace the unsafe version with the safe one.

Thanks for flagging it!

**Comment by Rinzii** [comment]
> I have an issue with this function:
> 
> ```c++
>     /**
>      * @brief Determines if a signed integer value is a power of two.
>      *
>      * @tparam T Type of the input.
>      * @param n Value to check.
>      * @return True if the input is a power of two, false otherwise.
>      *
>      * @note Prefer ccm::ext::safe::is_power_of_two for signed inputs to avoid signed
>      * overflow on the minimum representable value.
>      */
>     template <typename T, std::enable_if_t<std::is_signed_v<T> && !std::is_same_v<T, bool>, bool> = true>
>     constexpr bool is_power_of_two(T n) noexcept
>     { return n && !(n & (n - T(1))); }
> ```
> 
> inside is_power_of_two.hpp file. Because naturally negative numbers cannot be power of two. Therefore you should add `n > T(0)`
> 
> Another issue being that for functions `delta_angle`, `lerp_angle`, and `move_towards_angle` inside the namespace `ccm::ext` you use degrees but everywhere else I can see you're using radians. What if user forgets to apply radians to degrees? I think you should add a suffix `_deg` to tell the user those functions operate on degrees rather than radians or rewrite those functions to use radians.

Also fair point! These angle helpers borrow ideas from Unity and GLSL so that is where the original idea came from, but I agree that taking the approach of mirroring the rest of CCMath and the Standard by using radians is the right call here.

So I've decided we will key these off radians rather than add a _deg suffix. delta_angle, lerp_angle, and move_towards_angle now take and return radians, which keeps them in line with the trig functions. Also, while I was at it I decided to I flip the ext guard convention so the guarded behavior is now the default and the raw form lives under ccm::ext::unsafe rather than the opposite like before. I also went through the codegen on each unsafe variant and removed all the versions that did not produce better codegen.

**Review by Eczbek — APPROVED** [review]

**Review by nekoflare — APPROVED** [review]

File: docs/approximating_functions/sollya/sin_small_error.sollya [hunk 18]
```diff
@@ -0,0 +1,73 @@
+// Sampled error curve for the teaching sin_small kernel.
```
Comment (ProfessionalMenace) [inline docs/approximating_functions/sollya/sin_small_error.sollya:1]: some of the sollya script don't feel documented well enough

File: include/ccmath/internal/math/generic/func/basic/remquo_impl/remquo_double_impl.hpp [hunk 19]
```diff
@@ -42,7 +42,10 @@ namespace ccm::internal
 			y_i64 &= 0x7fffffffffffffffULL;
 
 			// If y is zero.
-			if (CCM_UNLIKELY(y_i64 == 0)) { return (x * y) / (x * y); }
+			if (CCM_UNLIKELY(y_i64 == 0))
+			{
+				return (x * y) / (x * y);
```
Comment (ProfessionalMenace) [inline include/ccmath/internal/math/generic/func/basic/remquo_impl/remquo_double_impl.hpp:51]: is this a good idea?

File: include/ccmath/internal/math/runtime/func/fmanip/ilogb_rt.hpp [hunk 20]
```diff
@@ -10,24 +10,22 @@
 
 #pragma once
 
+#include "ccmath/internal/math/generic/builtins/fmanip/ilogb.hpp"
```
Comment (ProfessionalMenace) [inline include/ccmath/internal/math/runtime/func/fmanip/ilogb_rt.hpp:13]: Seeing these imports more often makes me think that these header files should not be aware of the whole paths and that it should be done with a CMake/linker

**Review by ProfessionalMenace — APPROVED** [review]
Found nothing major. Refactoring fixed a lot of the readability issues I had.

File: docs/approximating_functions/sollya/sin_small_error.sollya (hunk 18 above)
Comment (Rinzii) [reply docs/approximating_functions/sollya/sin_small_error.sollya:1]: Fair though for now I'll treat this as deferred to a later date.

File: include/ccmath/internal/math/generic/func/basic/remquo_impl/remquo_double_impl.hpp (hunk 19 above)
Comment (Rinzii) [reply include/ccmath/internal/math/generic/func/basic/remquo_impl/remquo_double_impl.hpp:51]: Yes this is intentional. It's a common libm trick to produce a `NaN` and raise `FE_INVALID` at the same time without hard coding a `NaN` or pulling in `feraiseexception`. 

The idea is for `y == 0` it computes `x * 0` then `0 / 0`, which gives you the `NaN` and sets the invalid flag. It also properly handles when `x` is infinite. It is the same idiom used a few lines down for the non-finite case. I will though say I prob should of better documented that fact here. I'll add a comment in before merging.

File: include/ccmath/internal/math/runtime/func/fmanip/ilogb_rt.hpp (hunk 20 above)
Comment (Rinzii) [reply include/ccmath/internal/math/runtime/func/fmanip/ilogb_rt.hpp:13]: I see where your coming from though I will say for now this is a WONTFIX.
